Implement CI/CD GitHub Workflow

mchlopecki opened the pull request:
Addresses #7 

The workflow triggers on PRs inits and changes. The workflow can also be initiated from the [Actions tab](https://github.com/dmlary/godot-hex-map/actions), but only once pulled into main, and it will currently only run on the main branch in this implementation. 

The workflow attempts to build for Windows, Mac, and Ubuntu, but not Android and iOS. This does possibly leave a possibility for users of the extension to see errors for the Android editor/mobile exports. I think until the rewrite happens, the current builds will be sufficient. For Windows, I feel fairly confident in my os image choice, but I think the Mac/Ubuntu images could be checked. Notably, when referring to the FMOD GDExtension, the Ubuntu image was slated as deprecated, so it needed to be changed.

Some development notes are on [my fork](https://github.com/mchlopecki/godot-hex-map/pull/1).

mchlopecki commented:
Alright, ubuntu-latest is used now, and I added a `scons run_tests` at the end of the compile action. Should be ready to go. I forgot to mention that when I initially ran the workflow on my branch, the ubuntu builds were failing, and now adding the tests is making windows fail as well (looks like I forgot to run tests on my own compilations). Now, thankfully, you'll have the full logs to browse. Most recent run on my branch has all the details.

dmlary commented:
Also, do we need to run everything on `workflow_dispatch`, or was that just for debugging?

mchlopecki commented:
> Ok, let's do this, remove the tests from all the tasks. Add an independent task that runs the tests only on macos. I'll merge that PR, then start working at fixes to allow the tests to run on the other platforms.

I have this nearly ready to push, but I took another look at the FMOD workflow, and they only run tests on the `editor` target. That makes sense and we could reduce some run times on other builds if we think testing the editor build is sufficiently comprehensive. What do you think?

> Also, do we need to run everything on `workflow_dispatch`, or was that just for debugging?

From what I understand, `workflow_dispatch` allows the build test to be manually run on `main`, and before I realized every change to a PR would run the tests, I was planning on having later adding options to be able to run it on a specific branch. It's not strictly necessary anymore, though in the back of my mind I'm wondering about how to not have it run on every PR update in case you just make a naming/formatting change. It shouldn't incur any additional automatic runs, but I can still remove it.

dmlary commented:
> we could reduce some run times on other builds if we think testing the editor build is sufficiently comprehensive. What do you think?

Agreed, tests should just run on editor until we find a reason that's not sufficient.

> though in the back of my mind I'm wondering about how to not have it run on every PR update in case you just make a naming/formatting change.

I believe these lines will stop in-progress CI jobs when you push a new commit before it completes: https://github.com/cisco-open/ruby-ctypes/blob/9ed9858b45b2adee1e6b6503775a4ce098b7f9cb/.github/workflows/main.yml#L11-L13

mchlopecki commented:
Alright, along with now only running tests on mac/editor, I pulled the tests out to the main workflow to be able to more easily check and see whether the build failed or tests failed. I also added what I believe to be sufficient license information to the GitHub action that I borrowed from the FMOD repo, and added the cancelation of in-progress jobs when new commits are added. Once again, you can see how tests ran on [my fork](https://github.com/mchlopecki/godot-hex-map/pull/1).

dmlary commented:
Thanks for adding this!  It's going to be a huge help ensuring the extension works for all platforms.
